### DISCUSSION TITLE
Is it possible to read changelogs from github releases for npm packages?

### DISCUSSION THREAD

**Discussion by noahnu** [renovatebot/renovate#10746] [Request Help]
I have a private monorepo which publishes npm packages to a private registry. We don't store release notes in the repository, however we do have release notes in GitHub releases.

Renovate doesn't pick up the release notes from GitHub releases. Taking a look at the code https://github.com/renovatebot/renovate/blob/7eb2c09b6e465c45a3461e9237e4212bba9be132/lib/manager/npm/extract/index.ts#L187, it looks like "npm" is hardcoded to the npm datasource, whereas github releases requires the github-releases datasource.

Is that any way to tell renovate to read release notes from GitHub releases for the npm packages?

For the structure of the monorepo, it's identical to https://github.com/tophat/monodeploy, except without the CHANGELOG.md files in the packages directory -- everything else is identical.

---

Edit: Actually reading further, the logic is in `lib/workers/pr/changelog/index.ts`. When I check my logs for the job, it gets to the line "Fetching changelog: ${sourceUrl} (...". The source URL and source directory are correct, and I don't see any errors, yet the changelog  is not fetched from github releases.

An example tag for a release: `@private/my-package@0.4.7` where the dependency is `@private/my-package`.

I do see "Removing duplicate release notes" in the logs. Looking at https://github.com/renovatebot/renovate/blob/34ef699b64b5de1c6f9fe70eafbb99422b839081/lib/workers/pr/index.ts#L319, it seems to be keyed on the sourceUrl? If I have 2 packages from the same monorepo, they'll have the same source url but different source directory. These should be treated as separate release notes.

**Comment by rarkins** [comment]
Renovate already reads GitHub releases regardless if what the original datasource is, as long as the sourceUrl points to a GitHub repo. Possibly the tagging/naming of the releases is not one which Renovate detects correctly.

**Reply by noahnu** [reply]
The format is `@scope/package@1.2.3` which is pretty standard.

Could it be the duplicate release note logic that's causing the problem as I mentioned?

I'll try step through the code with a debugger.

**Reply by viceice** [reply]
Can be, feel free to open a issue / pr with some test cases / reproduction repo / fix, when you've verified the issue.

**Reply by noahnu** [reply]
Trying to debug this locally and I need to override the remote repository's config with a local config (to get around the missing private key). Is there a CLI option to specify a local config to take precedence?

**Reply by viceice** [reply]
try `force`

https://docs.renovatebot.com/self-hosted-configuration/#force

**Reply by noahnu** [reply]
Looks like that's not a CLI option and only for the config file which I can't define locally? It always uses the config file in the remote

I suppose I can just modify the code and hardcode my registry token

**Reply by viceice** [reply]
add a `config.js` file to your current directory where you start local renovate.

```js
module.exports = {
    force : {
        ....
    }
}
```

**Reply by noahnu** [reply]
I've confirmed it's picking up the release notes correctly in the getReleaseNotes function. So it must be filtering them out before it makes the PR somewhere. I still suspect it's something to do with the duplicates, because this is what one of my PRs looks like:

![image](https://user-images.githubusercontent.com/1297096/124791344-f99e2700-df19-11eb-9008-ce5d27ac1794.png)

The version link under 0.8.2 points to a package in the monorepo that's not included in the list above. That's why I think it must be deduplicating somehow, and using the first package it encounters in the monorepo even though that's not relevant.

**Reply by rarkins** [reply]
Yes, I do recall that there's some de-duplication that occurs. It maybe needs to be made more granular

**Reply by noahnu** [reply]
I have been able to confirm that the release notes only work correctly when there is one package from the monorepo. If there are 2 or more, only one of the release notes is displayed -- and it doesn't tell you which one is displayed, since it collapses it under the monorepo's name, rather than the package name.
